### PR TITLE
Telemetry: add the reason why auth token wasn't obtained

### PR DESCRIPTION
**Pull request by imakhalova**
Let's add property why auth token wasn't obtained

### CODE REVIEW

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 1]
```diff
@@ -1453,17 +1454,22 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
...
         // refresh token does not exist or refresh token request failed to give back the token. 
         // If it's non-silent request, will try to acquire token interactively. 
         // if it's silent request, will return the AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED.
         if (refreshItem == null || authResult == null 
                 || (authResult != null && StringExtensions.IsNullOrBlank(authResult.getAccessToken()))) {
             Logger.v(TAG, "Refresh token is not available or refresh token request failed to return token.");
+            Map<String, String> eventProperties = new InstrumentationPropertiesBuilder(request, authResult)
+                    .add(InstrumentationIDs.ERROR_CLASS, refreshItem == null ?
+                            InstrumentationIDs.REFRESH_TOKEN_NOT_FOUND :
+                            authResult == null ? InstrumentationIDs.AUTH_RESULT_EMPTY : InstrumentationIDs.AUTH_RESULT_CONTAINS_ERROR)
```
Comment (sdabbiru) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: AUTH_RESULT_CONTAINS_ERROR cannot be assumed. In this case, authResult does not contain an access token.

Comment (imakhalova) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: ok, let's rename it to NO_ACCESS_TOKEN_IN_RESPONSE. how does it sound?

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 2]
```diff
@@ -1396,6 +1397,14 @@ private AuthenticationResult acquireTokenAfterValidation(CallbackHandler callbac
                             ADALError.DEVELOPER_ACTIVITY_IS_NOT_RESOLVED));
                 }
             } else {
+                Map<String, String> eventProperties = new InstrumentationPropertiesBuilder(request, result)
+                        .add(InstrumentationIDs.IS_BROKER_APP, Boolean.TRUE.toString())
+                        .add(InstrumentationIDs.ERROR_CLASS,
+                                result == null ? InstrumentationIDs.AUTH_RESULT_EMPTY : InstrumentationIDs.NO_ACCESS_TOKEN_IN_RESPONSE)
+                        .build();
+                ClientAnalytics.logEvent(
+                        InstrumentationIDs.AUTH_TOKEN_NOT_RETURNED,
```
Comment (agamemnum) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: Irina that would be hard to use - let's use event name "TokenRequest" and two properties "Result":"Failed", "Reason":"AUTH_TOKEN_NOT_RETURNED". 

Overall do not create - event type for each condition. Instead design event types around scenario.

File: src/src/com/microsoft/aad/adal/InstrumentationIDs.java [hunk 3]
```diff
@@ -34,6 +34,11 @@
     /* Event invoked when auth token was successfully obtained */
     static final String REFRESH_TOKEN_REQUEST_SUCCEEDED = "RefreshTokenRequestSucceeded";
 
+    /* Error message that notifies that auth token wasn't obtained because there was no refresh token found*/
+    static final String REFRESH_TOKEN_NOT_FOUND = "RefreshTokenNotFound";
+    static final String AUTH_RESULT_EMPTY = "EmptyResponseFromServer";
+    static final String NO_ACCESS_TOKEN_IN_RESPONSE = "HasResponseFromServer";
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/InstrumentationIDs.java]: hm, NO_ACCESS_TOKEN_IN_RESPONSE = hasResponseFromServer? :)

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java (hunk 2 above)
Comment (imakhalova) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: sounds reasonable, changing

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 4]
```diff
@@ -31,6 +31,7 @@
 import java.security.NoSuchAlgorithmException;
 import java.util.Date;
 import java.util.Iterator;
+import java.util.Map;
 import java.util.UUID;
```
Comment (addev-ashish) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: this import seems to be redundant now

**Comment by addev-ashish** [comment]
Is there an issue filed for this PR?

**Comment by addev-ashish** [comment]
:clock1:

**Comment by imakhalova** [comment]
The issue that users complain that they got prompt for password too frequently and we need a way to collect statistics why this happens.

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java (hunk 4 above)
Comment (imakhalova) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: right, removing

**Comment by addev-ashish** [comment]
If there is no issue filed yet, i think its ok to leave this one as is. Going forward we want to keep a mapping between PR's and issues, so that we know the context in which the fixes were made in.

**Comment by weijjia** [comment]
:shipit:

**Pull request merged**